Make supportutils-plugin-ses cephadm aware (jsc#SES-1314)

tserong opened the pull request:
This change makes supportutils-plugin-ses use cephadm to get cluster and daemon status. As currently implemented it will still fall back to just invoking `ceph` if the command exists (i.e. if ceph-common is installed). I could use others' opinions on this:

1) Should I keep this behaviour (still try `ceph` if `cephadm` isn't present), or should I just not bother doing anything at all if cephadm isn't present?
2) The current implementation relies on /etc/ceph/ceph.conf and the admin keyring existing on the host (without these, `cephadm shell` won't be able to talk to the cluster). AFAICT these files will exist on the master/admin node (as configured by ceph-salt), and on the bootstrap node (first mon), but won't exist on other nodes. Should I switch to trying to find a running mgr, and interrogating via `podman exec` as suggested by @trociny? That wouldn't make any difference for storage nodes, but would allow any mgr node to return cluster status. OTOH, it wouldn't work anymore on the salt-master/admin node.
3) I'm using `podman exec` to invoke `ceph daemon ...` to interrogate individual daemons, but since implementing that realised that I could alternately use `cephadm enter`. Should I change to the latter, or is `podman exec` OK?
4) We've got `ceph config dump` already. Do we also need `ceph config-key dump`, or is the former enough?
5) Is there anything else we should be capturing?

Note:
* The current implementation is, in general, IMO, somewhat messy (there's repeated calls to `cephadm ls`, for example). This is largely due to the functionality of the plugin having accreted over time. The whole thing could probably use some restructuring, but I'd like to leave that aside until we've got it doing the right things :-)
* The iSCSI, Ganesha, Prometheus and Grafana bits are still TBD.

EDIT 2020-05-29: The above description was accurate when this PR was initially created, but things have changed since then, e.g. we always use `cephadm` and don't try to fall back to bare `ceph` invocations, and the iSCSI, Ganesha, Prometheus and Grafana bits are all actually done now. For more details, see the other comments below, or review the comments in the commits.

tserong committed:
Update RPM list for SUSE Enterprise Storage 7

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Don't capture obsolete calamari, deepsea, openATTIC info

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Capture ceph-salt config and log (jsc#SES-1529)

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Use `cephadm logs` to capture ceph logs

This will capture logs for any containerized service managed
by cephadm, including e.g.: prometheus and grafana.

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Collect status of containerized ceph daemons

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Use `cephadm shell` to collect cluster status

This also adds `cephadm ls` output and switches to using
`which` to find various binaries rather than using explicit
paths.

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Adjust path to crash info for containers

Now that everything's in containers, crash info is one level
deeper in the directory hierarchy (/var/lib/ceph/$FSID/crash),
hence the shell glob introduced here.  The -d check is in case
somehow no crash directories exist, in which case we end up with
the literal string '/var/lib/ceph/*/crash', which obviously won't
work when passed to `cp`.

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
List available and running container images

Signed-off-by: Tim Serong <tserong@suse.com>

sebastian-philipp commented on ses:
```diff
@@ -46,133 +46,156 @@ section_header "Ceph cluster status"
+
+if [ -x /usr/bin/ceph -o -n "${CEPHADM_SHELL}" ]; then
+    if [ -f /etc/ceph/ceph.conf ]; then
+        # None of the following will actually work without ceph.conf present on the host...
```
cephadm also searches for the ceph.conf in other locations: 

https://github.com/ceph/ceph/blob/8aac5d278525c45468d753c3542a820226019dda/src/cephadm/cephadm#L1082-L1110

I think for cephadm, we have to tweak the logic here a bit. Maybe of something like this?

```suggestion
    if [ $(${CEPHADM_SHELL} ceph ${ID} --connect-timeout=$CT -s > /dev/null) ]; then
        # None of the following will actually work without ceph.conf present on the host...
```

tserong replied:
Done in fc53c5f

sebastian-philipp commented on ses:
```diff
@@ -46,133 +46,156 @@ section_header "Ceph cluster status"
-                > $LOG/ceph/images/${pool}/rbd-info-${image} 2>&1
+        # timeout status commands after 5 seconds
+        CT=5
+        plugin_command "${CEPHADM_SHELL} ceph ${ID} --connect-timeout=$CT versions" > $LOG/ceph/ceph-versions 2>&1
```
Hm, instead of adding `CEPHADM_SHELL`, what about using
```suggestion
        plugin_command "${CEPH_CLI} ${ID}  versions" > $LOG/ceph/ceph-versions 2>&1
```
where CEPH_CLI is either

```bash
CEPH_CLI="cephadm shell -- ceph --connect-timeout=$CT"
```
or
```bash
CEPH_CLI="ceph --connect-timeout=$CT"
```

tserong replied:
We can't quite do that... I would work for all the `ceph` invocations, but there's a couple of cases where we run `rados` and `rbd`, not `ceph`, so I've left it as-is for now.

sebastian-philipp commented on ses:
```diff
@@ -46,133 +46,156 @@ section_header "Ceph cluster status"
-    fi
+if [ -x /usr/sbin/cephadm ] ; then
+    CEPHADM_SHELL="/usr/sbin/cephadm shell --"
+    plugin_command "cephadm ls" > $LOG/ceph/cephadm-ls 2>&1
```
I'd also add `cephadm check-host` and `cephadm inspect-image`

tserong replied:
Added in 55c2363

sebastian-philipp commented on ses:
```diff
@@ -46,133 +46,156 @@ section_header "Ceph cluster status"
+
+            mkdir -p $LOG/ceph/$daemon
+
+            podman exec ${container_id} ceph --connect-timeout=$CT daemon ${daemon} config show \
```
If this doesn't work:

```suggestion
            cephadm enter --name ${daemon}  --  ceph --connect-timeout=$CT daemon ${daemon} config show \
```

we should probably fix this in cehadm itself.

tserong replied:
I've change it to use `cephadm enter`, seems to work fine.

sebastian-philipp commented on ses:
```diff
@@ -194,62 +217,55 @@ if [ -d /etc/ceph ]; then
-    cp /var/log/calamari/*.log $LOG/ceph/log/calamari/
-    plugin_message "Calamari logs copied to ceph/log subdirectory"
+if [ -x /usr/sbin/cephadm ]; then
+    cephadm ls | jq -r '.[] | "\(.fsid) \(.name)"' |
+    while read fsid name ; do
+        mkdir -p $LOG/ceph/log/ceph/$fsid
+        # Individual daemon logs from journalctl named without the leading "ceph-"
+        # string to avoid any conflict with possible log files already copied.
+        plugin_command cephadm logs --fsid $fsid --name $name > $LOG/ceph/log/ceph/$fsid/$name.log 2>&1
+    done
```
Just created https://tracker.ceph.com/issues/45576

sebastian-philipp commented on ses:
```diff
@@ -339,7 +327,14 @@ fi
+if [ -x /usr/sbin/cephadm ]; then
+    cephadm ls | jq -r '.[] | "\(.fsid) \(.name)"' |
+    while read fsid name ; do
+        plugin_command "systemctl status -l 'ceph-$fsid@$name.service'"
```
looks odd to see `$fsid` enclosed within `'` chars (as in `status -l '$fsid'`). (Still, looks like this should work)

tserong replied:
Fixed.

sebastian-philipp commented on ses:
```diff
@@ -339,7 +327,14 @@ fi
 section_header "Ceph related services"
 
-plugin_command "systemctl status -l 'ceph*'"
+if [ -x /usr/sbin/cephadm ]; then
+    cephadm ls | jq -r '.[] | "\(.fsid) \(.name)"' |
```
looks odd to first query for `/usr/sbin/cephadm` but then use `cepham` to execute the commands. Maybe make them symmetric? Either call `$(which cephadm)` or run `/usr/sbin/cephadm ls`?

tserong replied:
Fixed in 285747b

sebastian-philipp commented on ses:
```diff
@@ -46,133 +46,156 @@ section_header "Ceph cluster status"
+        plugin_command "${CEPHADM_SHELL} timeout $CT rados ${ID} df" > $LOG/ceph/rados-df 2>&1
+        plugin_command "${CEPHADM_SHELL} ceph ${ID} --connect-timeout=$CT telemetry status" > $LOG/ceph/ceph-telemetry-status 2>&1
+        plugin_command "${CEPHADM_SHELL} ceph ${ID} --connect-timeout=$CT balancer status" > $LOG/ceph/ceph-balancer-status 2>&1
+
```
```bash
ceph orch status >> $LOG/ceph/cephadm-status 
ceph cephadm get-ssh-config >> $LOG/ceph/cephadm-status 
ceph orch ls --format yaml >> $LOG/ceph/cephadm-status 
ceph orch host ls --format yaml >> $LOG/ceph/cephadm-status 
ceph orch ps --format yaml >> $LOG/ceph/cephadm-status 
ceph orch device ls --format yaml >> $LOG/ceph/cephadm-status
ceph orch orch upgrade status >> $LOG/ceph/cephadm-status
```

tserong replied:
Added (although I called the log file `ceph-orch-status)

denisok commented on ses:
```diff
@@ -46,133 +46,156 @@ section_header "Ceph cluster status"
-                 awk -F. '{print "--id=" $3; exit}')
-    fi
+if [ -x /usr/sbin/cephadm ] ; then
+    CEPHADM_SHELL="/usr/sbin/cephadm shell --"
```
could we do it more general? Like CEPH_SHELL ?
I could see we could add something like
`CEPH_SHELL="kubectl -n rook-ceph exec -it $(kubectl -n rook-ceph get pod -l "app=rook-ceph-tools" -o jsonpath='{.items[0].metadata.name}') --"`
which makes same thing for the Rook

denisok replied:
but we would need another var to identify what env we are working in.
`-x /usr/sbin/cephadm ` and etc might just become bool vars...

tserong replied:
Done in f025955 (there's no actual rook support yet, but at least the name is generic now, to support this).

trociny commented on ses:
```diff
@@ -201,10 +201,24 @@ section_header "Ceph log files"
+        mkdir -p $LOG/ceph/log/ceph/$fsid
+        # Individual daemon logs from journalctl named without the leading "ceph-"
+        # string to avoid any conflict with possible log files already copied.
+        plugin_command cephadm logs --fsid $fsid --name $name > $LOG/ceph/log/ceph/$fsid/$name.log 2>&1
```
Can we avoid `$fsid` in the log path name? It just makes browsing the archive and processing with scripts harder.

I suppose the supportconfig operates in the assumption that the node is used for only one cluster. Otherwise we would need to tweak much more things to make the supportconfig support multicluster:  all commands like `ceph status` or `ceph daemon` should be per cluster then logging its output in the corresponding `$fsid` directory.

tserong replied:
I put `$fsid` in the log path there, because supportconfig is running on the VM host, and when in runs `cp -r /var/log/ceph/*` a few lines above, to collect any files that may be present, this already includes the FSID. So I kept it consistent with that.

You're right though, everything else assumes there's only one cluster, so having the FSID here in the log path is a bit... strange... from that perspective.

I guess the thing to do is get rid of FSID here, and also remove it when running `cp -r`?

The other alternative is of course to make supportconfig multi-cluster aware and have *everything* in an FSID path. Is there any sense in doing that work? (I'm leaning towards "no", because we don't expect people to run multiple clusters on the same hardware, but would welcome other opinions)

jan--f replied:
Not sure what the roadmap for multiple clusters is, maybe putting everything under an fsid folder is acceptable? So we'd get supportconfig with a dir tree like `ceph/<fsid>/...`?

trociny replied:
I think making it multi-cluster aware is a bit of work and out of scope here. For this PR I would just remove fsid from all log paths. If we eventually decide the multi-cluster support, it will need tweaking anyway.

tserong replied:
Agreed. 0e675c7 drops the FSID from the log directories.

tserong committed:
Collect ceph orchestrator status output

This adds the following:
- ceph orch status
- ceph cephadm get-ssh-config
- ceph orch ls
- ceph orch host ls
- ceph orch ps
- ceph orch device ls
- ceph orch upgrade status

Co-authored-by: Sebastian Wagner <swagner@suse.com>
Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Add cephadm check-host and inspect-image

Signed-off-by: Tim Serong <tserong@suse.com>

jan--f commented on ses:
```diff
@@ -46,133 +46,171 @@ section_header "Ceph cluster status"
-            plugin_command "timeout $CT rbd ${ID} -p $pool info $image" \
-                > $LOG/ceph/images/${pool}/rbd-info-${image} 2>&1
+
+    plugin_command "${CEPH_SHELL} ceph ${ID} --connect-timeout=$CT -s" > $LOG/ceph/ceph-status 2>&1
```
What happens here (and below) if `$CEPH_SHELL` is undefined? Its only set if cephadm is present in [line 49](https://github.com/SUSE/supportutils-plugin-ses/pull/33/files#diff-f2b32bda85a5a4a613eb47fb01c57ce3R49).

tserong replied:
It'll just invoke `ceph -s`. If that fails, the subsequent `if` will skip everything else.

jan--f replied:
OH duh...of course. Caffeine is kicking in now.

jan--f commented on ses:
```diff
@@ -175,14 +175,21 @@ else
     plugin_command "ceph-volume inventory --format json-pretty" > $LOG/ceph/ceph-volume-inventory.json 2>&1
     plugin_command "ceph-volume inventory" > $LOG/ceph/ceph-volume-inventory 2>&1
+elif [ -x /usr/sbin/ceph-disk ]; then
+    plugin_command "ceph-disk list" > $LOG/ceph/ceph-disk-list 2>&1
```
As I mentioned, no idea how the versioning works, but might be worth dropping `ceph-disk` here? mimic was the last release to ship `ceph-disk`

jan--f reviewed (commented):
Just a few thoughts here:

> Should I keep this behaviour (still try ceph if cephadm isn't present), or should I just not bother doing anything at all if cephadm isn't present?

Is there a point to keep the `ceph` behaviour? Not sure how supportconfig versioning works, but assuming newer version won't be backported to, say SLE15SP1 or earlier, we can probably drop the naked calls to ceph?

> I'm using podman exec to invoke ceph daemon ... to interrogate individual daemons, but since implementing that realised that I could alternately use cephadm enter. Should I change to the latter, or is podman exec OK?

I'm kinda leaning to `podman exec` if only for the potential of a bug in cephadm that we want to debug...?

> Is there anything else we should be capturing?

How about interrogating various podman commands? Just from looking at the podman docs there seems to be potentially interesting info to be had (though probably not all listed are interesting and I didn't list everything that is interesting):

-  http://docs.podman.io/en/latest/markdown/podman-history.1.html
- http://docs.podman.io/en/latest/markdown/podman-info.1.html
- http://docs.podman.io/en/latest/markdown/podman-inspect.1.html
- http://docs.podman.io/en/latest/markdown/podman-network-inspect.1.html
- http://docs.podman.io/en/latest/markdown/podman-port.1.html
- http://docs.podman.io/en/latest/markdown/podman-stats.1.html

Looking at LVM outside of ceph-volume might also be worthwhile.

Maybe its also worth looking at what info we can pull from the container runtime? ceph-salt has maybe interesting info to provide?

tserong committed:
Collect containerized NFS ganesha config

Logs and service status are already captured by `cephadm logs` and
`systemctl status -l ceph-$fsid@$name.service` invocations.

Signed-off-by: Tim Serong <tserong@suse.com>

trociny commented:
> Is there a point to keep the ceph behaviour? Not sure how supportconfig versioning works, but assuming newer version won't be backported to, say SLE15SP1 or earlier, we can probably drop the naked calls to ceph?

+1

tserong commented:
> > Is there a point to keep the ceph behaviour? Not sure how supportconfig versioning works, but assuming newer version won't be backported to, say SLE15SP1 or earlier, we can probably drop the naked calls to ceph?
> 
> +1

Done in f3925d8. This commit also drops the bare ceph-volume and ceph-disk invocations.

tserong committed:
Collect containerized SCSI Gateway config

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Collect size of containerized prometheus metrics

Signed-off-by: Tim Serong <tserong@suse.com>

tserong committed:
Don't try to use non-admin keyring for getting ceph status

In order to invoke all the `ceph` commands to get cluster
status, the current implementation tries to use whatever
keyring it can find in /etc/ceph/ if the admin keyring isn't
present.  Now that everything is running in containers, I
think it's less likely that any alternative keys will be
available to use on the host, so I've removed this
functionality.  This reverts 59843bd.

Note:
- The alternate keyring functionality wasn't working anyway
  with `cephadm shell`.  If we wanted to keep it, we'd have
  to add the --keyring paramenter to make cephadm use the
  right keyring, as well as setting $ID.
- Even without this functionality, supportconfig will still
  capture individual daemon state via `cephadm enter`.  It
  just won't get general cluster status.
- There will always be at least one node with an admin
  keyring (the bootstrap node), so cluster status can always
  be obtained by running supportconfig on that node.

Signed-off-by: Tim Serong <tserong@suse.com>

trociny commented on ses:
```diff
@@ -254,19 +254,21 @@ section_header "Ceph log files"
+    cp /var/log/ceph/*.log $LOG/ceph/log/ceph/
+    # Copy any log files in $FSID subdirectories (which is where containerized
+    # daemons may log to, at least for OSD creation and ceph-volume invocations)
+    cp /var/log/ceph/*/*.log $LOG/ceph/log/ceph/
```
what about copying compressed logs?

tserong replied:
Good point, 2dbbf71 should take care of that.

tserong commented:
OK, I'm reasonably happy with this now.

@jan--f, I agree it's probably worth looking at extra podman commands, LVM, etc. as you mentioned above, but I'd like to do those as separate issues, once this PR is merged.

@trociny, please let me know what you think of 06ebc0f, as it reverts the "try to use a non admin key" behaviour you introduced previously (more details are in the commit message).

trociny commented:
> @trociny, please let me know what you think of [06ebc0f](https://github.com/SUSE/supportutils-plugin-ses/commit/06ebc0f5dec80980eaf42dc67fa02e9fda1dac36), as it reverts the "try to use a non admin key" behaviour you introduced previously (more details are in the commit message).

I am fine with this as it would definitely not work for cephadm. I am not sure if we don't need something else instead. I am going to try the new version of the ses plugin the next week (it is going to be our squad SES7 testing week).

tserong commented:
OK, cool.

If nobody objects, I'll squash a few of these commits together, merge this PR then get an updated package built tomorrow.

denisok commented on ses:
```diff
@@ -46,133 +46,184 @@ section_header "Ceph cluster status"
 
+# TODO: for k8s/rook support, set CEPH_SHELL appropriately here
+
+if [ -n "${CEPH_SHELL}" ]; then
```
>  it will still fall back to just invoking ceph

but where does it fall back to the invoking ceph ? Or we don't push this to the SES6 ?
I don't see CEPH_SHELL is initialized to something default. Do we need that check at all ?

tserong replied:
CEPH_SHELL will be set if cephadm is present. If cephadm is not present, we won't be able to collect cluster status, but that's OK, because this change is only for SES7. Older releases live in other branches (ses6, ses5, ses4).

tserong commented:
> Looking at LVM outside of ceph-volume might also be worthwhile.

Supportconfig itself already grabs the following and dumps it in lvm.txt:

```
# /bin/systemctl status 'lvm2-activation-early.service'
# /sbin/pvs
# /sbin/vgs
# /sbin/lvs
# /sbin/dmsetup ls --tree
# /sbin/dmsetup table
# /sbin/dmsetup info
# /bin/ls -l --time-style=long-iso /etc/lvm/backup/
# /bin/ls -l --time-style=long-iso /etc/lvm/archive/
# /sbin/pvdisplay -vv
# /sbin/vgdisplay -vv
# /sbin/lvdisplay -vv
# /sbin/pvs -vvvv
# /sbin/pvscan -vvv
# /sbin/vgs -vvvv
# /sbin/lvs -vvvv
```
